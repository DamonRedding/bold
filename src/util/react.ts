import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
