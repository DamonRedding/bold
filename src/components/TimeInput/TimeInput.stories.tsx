import { action } from '@storybook/addon-actions'
import { boolean } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React from 'react'

import { TimeInput } from './TimeInput'

storiesOf('Components|TimeInput', module).add('default', () => (
  <TimeInput
    name='time'
    disabled={boolean('disabled', false)}
    guide={boolean('guide', true)}
    keepCharPositions={boolean('keepCharPositions', false)}
    onChange={action('changed')}
  />
))
