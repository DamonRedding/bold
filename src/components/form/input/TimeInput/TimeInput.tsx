import * as React from 'react'
import { conformToMask } from 'react-text-mask'
import createAutoCorrectedDatePipe from 'text-mask-addons/dist/createAutoCorrectedDatePipe'

import { Omit } from '../../../../util'
import { masks } from '../../../../util/masks'
import { MaskedInput, MaskedInputProps } from '../MaskedInput/MaskedInput'

export interface TimeInputProps extends Omit<MaskedInputProps, 'onChange'> {
    /**
     * Original input change event handler
     */
    onInputChange?: MaskedInputProps['onChange']

    onChange?(value: string): any
}

export class TimeInput extends React.PureComponent<TimeInputProps> {
    render() {
        const { onChange, ...rest } = this.props
        return (
            <MaskedInput
                mask={masks.time}
                placeholder='hh:mm'
                pipe={createAutoCorrectedDatePipe('HH:MM')}
                {...rest}
                onChange={this.handleChange}
                onBlur={this.handleBlur}
            />
        )
    }

    handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        if (!e) {
            return this.props.onChange(null)
        }
        if (this.props.onChange) {
            const value = e.target.value
            this.props.onChange(value)
        }
        if (this.props.onInputChange) {
            return this.props.onInputChange(e)
        }
    }

    handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
        if (this.props.onChange) {
            const value = e.target.value
            if (value) {
                const paddedValue = padTime(value)
                this.props.onChange(paddedValue)
            }
        }

        if (this.props.onBlur) {
            // Call original blur handler (if existent)
            return this.props.onBlur(e)
        }
    }
}

export const padTime = (value: string): string => {
    const numbersOnly = value.replace(/[^\d]/g, '')
    if (numbersOnly.length < 4) {
        const paddedValue = numbersOnly.padEnd(4, '0')
        const conformedValue = conformToMask(paddedValue, masks.time, {})
        return conformedValue.conformedValue
    }
    return value
}
