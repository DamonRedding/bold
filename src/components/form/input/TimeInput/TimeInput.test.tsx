import { shallow } from 'enzyme'
import * as React from 'react'

import { padTime, TimeInput } from './TimeInput'

describe('render', () => {
    it('should render correcly', () => {
        expect(shallow(<TimeInput name='time' />)).toMatchSnapshot()
    })
})

describe('onChange', () => {
    it('should be called with input value when input is changed', () => {
        const change = jest.fn()
        const wrapper = shallow(<TimeInput name='time' onChange={change} />)
        expect(change).not.toHaveBeenCalled()
        wrapper.simulate('change', { target: { value: '10:23' } })
        expect(change).toHaveBeenCalledWith('10:23')
    })
})

describe('onChange', () => {
    it('should be called with original input event when input is changed', () => {
        const inputChange = jest.fn()
        const wrapper = shallow(<TimeInput name='time' onInputChange={inputChange} />)
        expect(inputChange).not.toHaveBeenCalled()
        wrapper.simulate('change', { target: { value: '01:48' } })
        expect(inputChange).toHaveBeenCalledWith(expect.objectContaining({ target: { value: '01:48' } }))
    })
})

describe('onBlur', () => {
    it('should call the onChange prop with padded time', () => {
        const change = jest.fn()
        const wrapper = shallow(<TimeInput name='time' onChange={change} />)
        wrapper.simulate('blur', { target: { value: '1' } })
        expect(change).toHaveBeenCalledWith('10:00')
    })
    it('should call the original onBlur prop', () => {
        const blur = jest.fn()
        const wrapper = shallow(<TimeInput name='time' onBlur={blur} />)
        expect(blur).not.toHaveBeenCalled()
        wrapper.simulate('blur', { target: { value: '01:00' } })
        expect(blur).toHaveBeenCalled()
    })
})

describe('padTime', () => {
    it('should complete value with zeros', () => {
        expect(padTime('')).toEqual('00:00')
        expect(padTime('1')).toEqual('10:00')
        expect(padTime('13')).toEqual('13:00')
        expect(padTime('14:5')).toEqual('14:50')
        expect(padTime('23:56')).toEqual('23:56')
    })
})
