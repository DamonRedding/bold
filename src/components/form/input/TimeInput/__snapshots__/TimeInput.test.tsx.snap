// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render correcly 1`] = `
<MaskedInput
  mask={
    Array [
      /\\\\d/,
      /\\\\d/,
      ":",
      /\\\\d/,
      /\\\\d/,
    ]
  }
  name="time"
  onBlur={[Function]}
  onChange={[Function]}
  pipe={[Function]}
  placeholder="hh:mm"
/>
`;
